core/utils/utils.go

```diff
@@ -447,15 +453,27 @@ func CombinedContext(signals ...interface{}) (context.Context, context.CancelFun
 
 		switch sig := signal.(type) {
 		case context.Context:
-			ch = reflect.ValueOf(sig.Done())
+			if deadline, ok := sig.Deadline(); ok {
+				// Special handling for deadline contexts so we preserve the correct error
+				var deadlineCancel context.CancelFunc
+				ctx, deadlineCancel = context.WithDeadline(ctx, deadline)
+				cancels = append(cancels, deadlineCancel)
+
+				// Still need to handle the case where a deadline context is canceled
+				ch = reflect.ValueOf(sig.Done())
```

Comment:
Question to reviewer: can this race?

e.g. is it possible for the original deadline context to return and then we cancel this instead of getting a context.DeadlineExceeded?